### PR TITLE
Artifact movement refactoring

### PR DESCRIPTION
**Pull request by SoundSSGood**
This PR brings VCMI behavior closer to OH3
Fixed slot highlighting. Now the backpack slot is not highlighte.
backpack_highlighted
And other highlighting errors.
backpack_highlighted1

Fixed a bug while drag&drop swapping an equipped artifact (part of a collection). VCMI ask's to assemble the artifact if the rest of the parts were in the backpack.
Also now, when drag&drop artifacts swap, the artifact is moved to a special transit place instead of a backpack. This is a step towards the implementation the backpack size limit in the future.

### CODE REVIEW

File: client/widgets/CArtifactHolder.cpp [hunk 1]
```diff
@@ -278,13 +278,12 @@ void CHeroArtPlace::select ()
 	if (locked)
 		return;
 
-	selectSlot(true);
 	pickSlot(true);
 	if(ourArt->canBeDisassembled() && slotID < GameConstants::BACKPACK_START) //worn combined artifact -> locks have to disappear
 	{
-		for(int i = 0; i < GameConstants::BACKPACK_START; i++)
+		for(auto slot : ArtifactUtils::constituentWornSlots())
```
Comment (IvanSavenko) [inline client/widgets/CArtifactHolder.cpp:284]: Change of behavior: old code included war machines, new one does not. Intended?

File: client/widgets/CArtifactHolder.cpp [hunk 2]
```diff
@@ -309,9 +308,9 @@ void CHeroArtPlace::deselect ()
 	pickSlot(false);
 	if(ourArt && ourArt->canBeDisassembled()) //combined art returned to its slot -> restore locks
 	{
-		for(int i = 0; i < GameConstants::BACKPACK_START; i++)
+		for(auto slot : ArtifactUtils::constituentWornSlots())
```
Comment (IvanSavenko) [inline client/widgets/CArtifactHolder.cpp:311]: Also here, same change of behavior

File: client/windows/GUIClasses.cpp [hunk 3]
```diff
@@ -1249,8 +1249,6 @@ CExchangeWindow::CExchangeWindow(ObjectInstanceID hero1, ObjectInstanceID hero2,
 
 CExchangeWindow::~CExchangeWindow()
```
Comment (IvanSavenko) [inline client/windows/GUIClasses.cpp:1250]: Empty destructor - delete it entirely?

File: lib/CArtHandler.cpp [hunk 4]
```diff
@@ -851,15 +856,15 @@ void CArtifactInstance::putAt(ArtifactLocation al)
 	assert(canBePutAt(al));
 
 	al.getHolderArtSet()->setNewArtSlot(al.slot, this, false);
-	if(!ArtifactUtils::isSlotBackpack(al.slot))
+	if(!ArtifactUtils::isSlotBackpack(al.slot) && (al.slot != ArtifactPosition::TRANSITION_POS))
```
Comment (IvanSavenko) [inline lib/CArtHandler.cpp:859]: Perhaps add another method similar to isSlotBackpack, e.g. isSlotEquipment (which returns true for all slots that are considered as "equipped" and provide bonuses to hero)

File: client/widgets/CArtifactHolder.cpp (hunk 1 above)
Comment (SoundSSGood) [reply client/widgets/CArtifactHolder.cpp:284]: yes, intended

File: client/widgets/CArtifactHolder.cpp (hunk 2 above)
Comment (SoundSSGood) [reply client/widgets/CArtifactHolder.cpp:311]: and here too

File: client/windows/GUIClasses.cpp (hunk 3 above)
Comment (SoundSSGood) [reply client/windows/GUIClasses.cpp]: done

File: lib/CArtHandler.cpp (hunk 4 above)
Comment (SoundSSGood) [reply lib/CArtHandler.cpp]: done

**Pull request merged**